src/type.js

```diff
@@ -220,17 +225,28 @@ async function typeImpl(
           newEntry = `-${char}`
         } else if (prevWasPeriod) {
           newEntry = `${prevValue}.${char}`
+        } else if (isValidDateValue(currentElement(), textToBeTyped)) {
+          newEntry = textToBeTyped
         }
 
-        const inputEvent = fireInputEventIfNeeded({
-          ...calculateNewValue(newEntry, currentElement(), currentValue()),
-          eventOverrides: {
-            data: key,
-            inputType: 'insertText',
-            ...eventOverrides,
-          },
-        })
-        prevValue = inputEvent.prevValue
+        if (
+          !(currentElement().type === 'date') ||
+          isValidDateValue(currentElement(), textToBeTyped)
+        ) {
```

Comment:
[`date` input doesn't trigger an `input` event when a date is not valid.](https://codesandbox.io/s/user-event-playground-91jlf).

My first move was doing this conditional inside the `fireInputEventIfNeeded`, but the conditional there is more general, so I keep it here.